### PR TITLE
Improve getting `item class` in `ListComponent`

### PR DESCRIPTION
**Pull request by M1troll**
Restore ability to specify the `item_class` attribute in `ListComponent` inheritors.

This is because getting `item_class` from the first generic variable could become invalid when inheriting and adding new generic variables to the component.

Autocomplete `item_class` with the value passed in `Generic[ListItemType]` is still available. But it will be overridden by the `item_class` attribute if it is explicitly specified in the component class.

Examples:

## Case 1 - indication `item_class` in base class with generic
```python
from typing import Generic, TypeVar
from pomcorn import Component, ListComponent, Page, locators
from selenium import webdriver as selenium_webdriver

# pypi search page
url = "https://pypi.org/search/?q=saritasa&o="
# Set browser's language to English

options = selenium_webdriver.ChromeOptions()
prefs = {"intl.accept_languages": "en,en_U"}
options.add_experimental_option("prefs", prefs)

# Open page
webdriver = selenium_webdriver.Chrome(options)
page = Page.open(webdriver, app_root=url)

# Prepare out item class component
class ExpectedItemClass(Component[Page]):...

# Implement list component from which another component with generic variable will inherit
TPage = TypeVar("TPage")
class ListItem(Generic[TPage], ListComponent[ExpectedItemClass, TPage]):
    base_locator = locators.PropertyLocator(
        prop="aria-label",
        value="Search results",
    )
    relative_base_locator = locators.TagNameLocator("li")

# Create a descendant and pass a generic variable to it
class InheritedList(ListItem[Page]):...

# Check current `item_class`
print(f"`InheritedList.item_class` is `{InheritedList(page)._item_class}` now")
# `InheritedList._item_class` is `~TPage` now
# Instead of specified `<class '__main__.ExpectedItemClass'>`

# To solve this problem just specify ``item_class`` right in ``InheritedList``:

class InheritedList2(ListItem[Page]):
	item_class = ExpectedItemClass

# Check current `item_class`
print(f"`InheritedList2.item_class` is `{InheritedList2(page)._item_class}` now")
# `InheritedList.item_class` is `<class '__main__.ExpectedItemClass'>` now

webdriver.close()

```

## Case 2 - indication `item_class` in child class
```python
from typing import Generic, TypeVar, TypeAlias
from pomcorn import Component, ListComponent, Page, locators
from selenium import webdriver as selenium_webdriver

# pypi search page
url = "https://pypi.org/search/?q=saritasa&o="
# Set browser's language to English

options = selenium_webdriver.ChromeOptions()
prefs = {"intl.accept_languages": "en,en_U"}
options.add_experimental_option("prefs", prefs)

# Open page
webdriver = selenium_webdriver.Chrome(options)
page = Page.open(webdriver, app_root=url)

TRow = TypeVar("TRow")
TPage = TypeVar("TPage")


class Table(Generic[TRow, TPage], ListComponent[TRow, TPage]):
    base_locator = locators.PropertyLocator(
        prop="aria-label",
        value="Search results",
    )
    relative_base_locator = locators.TagNameLocator("li")


CostRow: TypeAlias = Component[Page]
class CostsTypeTable(Table[CostRow, Page]):
    base_locator = locators.PropertyLocator(
        prop="aria-label",
        value="Search results",
    )
    relative_base_locator = locators.TagNameLocator("li")


cls = CostsTypeTable(page)
print(f"{cls.item_class=}")
webdriver.close()

```

### CODE REVIEW

File: pomcorn/component.py [hunk 1]
```diff
@@ -192,21 +194,26 @@ def __init__(
...
-        if item_class := getattr(self, "item_class", None):
+        list_item_type = self._get_list_item_class()
+        item_class = getattr(self, "item_class", None)
+
+        if item_class:
+            self.item_class = item_class
+        else:
+            self.item_class = list_item_type
+
+        if item_class == list_item_type:
```
Comment (TheSuperiorStanislav) [inline pomcorn/component.py:205]: Suggest to move this logic to `__init_subclass__`

File: docs/CHANGELOG.rst [hunk 2]
```diff
@@ -3,6 +3,17 @@ Version history
 
 We follow `Semantic Versions <https://semver.org/>`_.
 
+0.8.1 (15.11.24)
+*******************************************************************************
+- Restore ability to specify ``item_class`` in ``ListComponent``. This is
+  because getting ``item_class`` from the first generic variable may become
+  invalid when inheriting and adding new generic variables in component.
```
Comment (TheSuperiorStanislav) [inline docs/CHANGELOG.rst:10]: Could you add an example of such a case? For future reference? Maybe even add it to docs?

File: pomcorn/component.py [hunk 3]
```diff
@@ -253,26 +270,27 @@ def all(self) -> list[ListItemType]:
...
             items.append(
-                self._item_class(page=self.page, base_locator=locator),
+                self.item_class(page=self.page, base_locator=locator),
             )
         return items
 
+    @classmethod
+    def get_list_item_class(cls) -> type[ListItemType]:
+        """Return class passed in `Generic[ListItemType]`."""
+        return get_args(cls.__orig_bases__[0])[0]  # type: ignore
```
Comment (TheSuperiorStanislav) [inline pomcorn/component.py:280]: Thanks to the example I see the problem. Could you try this?

```suggestion
    @classmethod
    def get_list_item_class(cls) -> type[ListItemType]:
        """Return class passed in `Generic[ListItemType]`."""
        base_class = next(
            klass
            for klass in cls.__orig_bases__  # type: ignore
            if inspect.isclass(typing.get_origin(klass))
            and issubclass(typing.get_origin(klass), ListComponent)
        )
        return get_args(base_class)[0]  # type: ignore
```

File: demo/pages/search_page/components/package_list.py [hunk 4]
```diff
@@ -5,12 +5,18 @@
 from .package import Package
 
 
+# The ``ListComponent`` item (Package) should always be ``Component``, because
```
Comment (ron8mcr) [inline demo/pages/search_page/components/package_list.py:8]: Looks like this comment is not actual now (since removing of components w/o base locators)

**Review by ron8mcr — APPROVED** [review]

File: pomcorn/component.py (hunk 3 above)
Comment (M1troll) [reply pomcorn/component.py:280]: Unfortunately I still get an wrong value on example code =(
image

`__orig_bases__` not store values of parent components
image

Comment (TheSuperiorStanislav) [reply pomcorn/component.py:280]: It works, but not in a way I expected to) 
image

**Comment by M1troll** [comment]
@TheSuperiorStanislav @ron8mcr 

I used suggestion from https://github.com/saritasa-nest/pomcorn/pull/98#discussion_r1843124675 by @TheSuperiorStanislav but its not cover all cases
So I collect some more cases where `item_class` should works and refactor `__init_subclass__`
Currectn implementation works fine with all of them

How do you think, maybe add this cases as tests?

## Common Part
```python
from typing import Generic, TypeVar
from pomcorn import Component, ListComponent, Page, locators
from selenium import webdriver as selenium_webdriver

# pypi search page
url = "https://pypi.org/search/?q=saritasa&o="
options = selenium_webdriver.ChromeOptions()
# Set browser's language to English
prefs = {"intl.accept_languages": "en,en_U"}
options.add_experimental_option("prefs", prefs)

# Open page
webdriver = selenium_webdriver.Chrome(options)
page = Page.open(webdriver, app_root=url)

TRow = TypeVar("TRow")
TPage = TypeVar("TPage")
class Item(Component[Page]):...
```

## Case 1 - item_class specified in parent class, but child contains another generic variable
```python
TPage = TypeVar("TPage")
class BaseList(Generic[TPage], ListComponent[Item, TPage]):
    base_locator = locators.TagNameLocator("body")

# Create a descendant and pass a generic variable to it
class InheritedList(BaseList[Page]):...

cls = InheritedList(page)
print(f"Case 1: {cls.item_class=}")
```

## Case 2 - If item class is not specified we should get an error
```python
class BaseList(Generic[TRow, TPage], ListComponent[TRow, TPage]):
    base_locator = locators.TagNameLocator("body")

class List(Generic[TRow], BaseList[TRow, Page]):
    base_locator = locators.TagNameLocator("body")

cls = List(page)
try:
    print(f"Case 2: {cls.item_class=}")
except ValueError as er:
    print(f"Case 2: {er}")
    pass
```

## Case 3 - item class specified only in child component
```python
class BaseList(Generic[TRow, TPage], ListComponent[TRow, TPage]):
    base_locator = locators.TagNameLocator("body")

class List(BaseList[Item, Page]):
    base_locator = locators.TagNameLocator("body")


cls = List(page)
print(f"Case 3: {cls.item_class=}")
```

## Case 4 - Valid case with all specified data
```python
class List(ListComponent[Item, Page]):
    base_locator = locators.TagNameLocator("body")

cls = List(page)
print(f"Case 4: {cls.item_class=}")
```

## Case 5 - New generic variable
```python
TParam = TypeVar("TParam")
class BaseList(Generic[TParam], ListComponent[Item, Page]):
    base_locator = locators.TagNameLocator("body")
    param: TParam | None = None

class Param(Component[Page]):...
class List(BaseList[Param]):
    base_locator = locators.TagNameLocator("body")

cls = List(page)
print(f"Case 5: {cls.item_class=}")
```

## Case 6 - Override item_class attr
```python
class List(ListComponent[Item, Page]):
    base_locator = locators.TagNameLocator("body")

class OverriddenItem(Component[Page]):...
class OverriddenList(List):
    item_class = OverriddenItem

cls = OverriddenList(page)
print(f"Case 6: {cls.item_class=}")
```

## Case 7 - Useless override of item_class (check warnings in __init_subclass__)
```python
class List(ListComponent[Item, Page]):
    item_class = Item
    base_locator = locators.TagNameLocator("body")

cls = List(page)
print(f"Case 7: {cls.item_class=}")
```

## Case 8 - Useless override of item_class (check warnings in item_class.setter)
```python
class List(ListComponent[Item, Page]):
    base_locator = locators.TagNameLocator("body")

cls = List(page)
cls.item_class = Item
print(f"Case 8: {cls.item_class=}")


webdriver.close()
```

**Review by ron8mcr — COMMENTED** [review]
> How do you think, maybe add this cases as tests?

Yes, the proposed logic is pretty complex and it's refactoring can be hard, so let's 
a. either simplify it like it was before (It looks to me like the previous implementation covered 90% of the cases with 10% of the effort) 
b. or add some kind of tests for it

**Comment by TheSuperiorStanislav** [comment]
> > How do you think, maybe add this cases as tests?
> 
> Yes, the proposed logic is pretty complex and it's refactoring can be hard, so let's a. either simplify it like it was before (It looks to me like the previous implementation covered 90% of the cases with 10% of the effort) b. or add some kind of tests for it

Not what I had in mind, the idea was to drop the need to specify item_class completliy(only via generics) and see if there would be a need for manual, try different cases. Curretly there is three things _item_class, item_class(this a property or attr) and our method for fething from generic args, I propose to just leave `fething from generic args` logic.

**Comment by M1troll** [comment]
@TheSuperiorStanislav 

> Not what I had in mind, the idea was to drop the need to specify item_class completliy(only via generics) and see if there would be a need for manual, try different cases. Curretly there is three things _item_class, item_class(this a property or attr) and our method for fething from generic args, I propose to just leave fething from generic args logic.

My bad, I forgot to add why I left the `item_class` attribute
Sometimes there are cases where a `ListComponent` contains multiple types of item classes. For example, a tree of files which include files and folders

And in that case I did the following - override `self.item_class`:
```python
    from pomcorn import Page, Component. ListComponent

    class FileSystemObj(Component[Page]):...
    class File(FileSystemObj):...
    class Folder(FileSystemObj):...

    class FilesTree(ListComponent[FileSystemObj, Page]):
        # Tree can contain files and folders, and to precisely specify
        # item_class component we set `item_class` right in methods below
        def get_folder(self, name: str) -> Folder:
            self.item_class = Folder
            return super().get_item_by_text(name)

        def get_file(self, name: str) -> File:
            self.item_class = File
            return super().get_item_by_text(name)

```

Now I realize I can just override `get_item_by_text`, so I removed the ability to override `item_class` as you wish

**Review by ron8mcr — APPROVED** [review]

**Review by NikAzanov — APPROVED** [review]

**Pull request merged**